### PR TITLE
persist: disable compaction in read-only mode

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR disables Persist's compaction when environmentd or clusterd are in read-only mode. As part of https://github.com/MaterializeInc/materialize/pull/30205 we discovered that during a 0dt deployment the read-only instance of Materialize was scheduling compaction requests which was causing it to write data.

We disable compaction by adding a `process_compaction_requests: Arc<AtomicBool>` to the `PersistConfig`, and when setting it to `true` when `clusterd` receives the already existing `AllowWrites` command. Also included is a CYA dyncfg that gates whether or not we check the flag, it's set to `true` by default.

I also added a Prometheus metric to count the number of requests dropped because compaction is disabled, a unit test to exercise the basic enable/disable behavior, and manually checked when running a 0dt test that compaction was successfully disabled and then enabled across `environmentd` and all `clusterd`s when a deployment was promoted.

### Motivation

Fix issue found in https://github.com/MaterializeInc/materialize/pull/30205

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/environmentd/src/lib.rs [hunk 1]
```diff
@@ -586,6 +585,13 @@ impl Listeners {
             (adapter_storage, audit_logs_handle)
         };
 
+        // Enable Persist compaction if we're not in read only.
+        if read_only {
+            config.controller.persist_clients.cfg().disable_compaction();
+        } else {
```
Comment (benesch) [inline src/environmentd/src/lib.rs:591]:
```suggestion
        if !read_only {
```

Comment (benesch) [reply src/environmentd/src/lib.rs:591]: I find it marginally clearer to rely on the fact that compaction stats out disabled!

**Review by benesch — COMMENTED** [review]
Nice! Deferring the real review to @bkirwi but no qualms from me.

File: src/persist-client/src/internal/compact.rs [hunk 2]
```diff
@@ -144,6 +144,13 @@ pub(crate) const COMPACTION_USE_MOST_RECENT_SCHEMA: Config<bool> = Config::new(
     ",
 );
 
+pub(crate) const COMPACTION_CHECK_PROCESS_FLAG: Config<bool> = Config::new(
+    "persist_compaction_check_process_flag",
+    true,
+    "CYA. Whether or not Compactor will check the process_requests flag in PersistConfig,\
+        which allows dynamically disabling compaction.",
```
Comment (bkirwi) [inline src/persist-client/src/internal/compact.rs:151]:
```suggestion
    "Whether Compactor will obey the process_requests flag in PersistConfig, \
        which allows dynamically disabling compaction. If false, all compaction requests will be processed.",
```

**Review by bkirwi — APPROVED** [review]
LGTM!

Made a suggestion on the doc to make it more clear what the default behaviour will be when the dyncfg is disabled.

File: src/persist-client/src/internal/compact.rs (hunk 2 above)
Comment (ParkMyCar) [reply src/persist-client/src/internal/compact.rs]: Makes sense to me!

**Pull request merged**